### ISSUE TITLE
Unable to launch container -  java null pointer errors

### ISSUE THREAD

**Issue by sudonem** [Schaka/janitorr#85]
Attempting to deploy Janitorr, but java related errors are preventing the container from even launching once. I concede that this might be my fault my configuration is pretty straight forward.

### General notes
- I created a new stack in portainer on a debian LXC for janitorr, though I'd eventually migrate it to my arr stack.
- The data is all located on a mapped nfs share: /nfs/STREAMING_MEDIA/data

### Docker compose file:
```
version: '3'

services:
  janitorr:
    container_name: janitorr
    image: ghcr.io/schaka/janitorr:stable
    volumes:
      - /config/janitorr/config/application.yml:/workspace/application.yml
      - /nfs/STREAMING_MEDIA/data:/data
    healthcheck:
      test: "wget -T5 -qO- http://localhost:8081/health | grep UP || exit 1"
      start_period: 30s
      interval: 5s
      retries: 3
```
(I've tried using native-stable, stable and develop - same result each time)

### application.yml file:
```
server:
  port: 8978

# File system access (same mapping as Sonarr, Radarr and Jellyfin) is required to delete TV shows by season and create "Leaving Soon" collections in Jellyfin
# Currently, Jellyfin does not support an easy way to add only a few seasons or movies to a collection, we need access to temporary symlinks
# Additionally, checks to prevent deletion on currently still seeding media currently require file system access as well
file-system:
  access: true
  validate-seeding: true # validates seeding by checking if the original file exists and skips deletion - turning this off will send a delete to the *arrs even if a torrent may still be active
  leaving-soon-dir: "/data/media/leaving-soon" # The directory that's known to Janitorr - this will contain new folders with symlinks to your media library for "Leaving Soon"
  media-server-leaving-soon-dir: "/data/media/leaving-soon" # This is the directory Jellyfin/Emby will be told it can find the "Leaving Soon" library, in case its mapped differently
  from-scratch: true # Clean up entire "Leaving Soon" directory and rebuild from scratch - this can help with clearing orphaned data - turning this off can save resources (less writes to drive)
  free-space-check-dir: "/" # This is the default directory Janitorr uses to check how much space is left on your drives. By default, it checks the entire root - you may point it at a specific folder

application:
  dry-run: true
  whole-tv-show: false # activating this will treat as a whole show as recently download/watched from a single episode, rather than that episode's season - shows will be deleted as a whole
  whole-show-seeding-check: false # Turning this off, disables the seeding check entirely if whole-tv-show is enabled. Activating this check will keep a whole TV show if any season is still seeding (requires file access).
  leaving-soon: 14d # 14 days before a movie is deleted, it gets added to a "Leaving Soon" type collection (i.e. movies that are 76 to 89 days old)
  exclusion-tag: "janitorr_keep" # Set this tag to your movies or TV shows in the *arrs to exclude media from being cleaned up

  media-deletion:
    enabled: true
    movie-expiration:
      # Percentage of free disk space to expiration time - if the highest given number is not reached, nothing will be deleted
      # If filesystem access is not given, disk percentage can't be determined. As a result, Janitorr will always choose the largest expiration time.
      5: 15d # 15 days
      10: 30d # 1 month - if a movie's files on your system are older than this, they will be deleted
      15: 30d # 2 months
      20: 90d # 3 months
    season-expiration:
      5: 15d # 15 days
      10: 20d # 20 days - if a season's files on your system are older than this, they will be deleted
      15: 60d # 2 months
      20: 120d # 4 months

  tag-based-deletion:
    enabled: true
    minimum-free-disk-percent: 100
    schedules:
      - tag: 5 - demo
        expiration: 30d
      - tag: 10 - demo
        expiration: 7d

  episode-deletion: # This ignores Jellystat. Only grab history matters. It also doesn't clean up Jellyfin. There is NO seeding check either.
    enabled: true
    tag: janitorr_daily # Shows tagged with this will have all episodes of their LATEST season deleted by the below thresholds
    max-episodes: 10 # maximum (latest) episodes of this season to keep
    max-age: 30d # Maximum age to keep any episode at all - even the last 10 episodes would expire after 30 days in this example


clients:
  sonarr:
    enabled: true
    url: "http://10.20.0.11:8989"
    api-key: "REDACTED"
    delete-empty-shows: true # Delete empty shows if deleting by season. Otherwise leaves Sonarr entries behind.
    determine-age-by: most_recent # Optional property, use 'most_recent' or 'oldest' - remove this line if Janitorr should determine by upgrades enabled for your profile
  radarr:
    enabled: true
    url: "http://10.20.0.11:7878"
    api-key: "REDACTED"
    only-delete-files: true # NOT RECOMMENDED - When set to true, Janitorr will only delete your media files but keep the entries in Radarr
    determine-age-by: most_recent # Optional property, use 'most_recent' or 'oldest' - remove this line if Janitorr should determine by upgrades enabled for your profile

  ## You can only choose one out of Jellyfin or Emby.
  ## User login is only needed if deletion is enabled.
  jellyfin:
    enabled: false
    url: "http://localhost:8096"
    api-key: "REDACTED"
    username: Janitorr
    password: janitorr
    delete: true # Jellyfin setup is required for JellyStat. However, if you don't want Janitorr to send delete requests to the Jellyfin API, disable it here
    leaving-soon-tv: "Shows (Leaving Soon)"
    leaving-soon-movies: "Movies (Leaving Soon)"

  emby:
    enabled: true
    url: "http://10.20.0.17:8096"
    api-key: "REDACTED"
    username: Janitorr
    password: REDACTED
    delete: true # Emby setup is required for JellyStat. However, if you don't want Janitorr to send delete requests to the Emby API, disable it here
    leaving-soon-tv: "Shows (Leaving Soon)"
    leaving-soon-movies: "Movies (Leaving Soon)"

  jellyseerr:
    enabled: true
    url: "http://10.40.0.10:5055"
    api-key: "REDACTED"
    match-server: false # Enable if you have several Radarr/Sonarr instances set up in Jellyseerr. Janitorr will match them by the host+port supplied in their respective config settings.

  jellystat:
    enabled: false
    whole-tv-show: false # Enabling this will make Jellystat consider TV shows as a whole if any episode of any season has been watched
    url: "http://jellystat:3000"
    api-key: "jellystat-key"
```

### Console error output.

```
 :: Spring Boot ::             (v3.4.0-M3)
2024-10-13T16:04:53.748Z  INFO 1 --- [           main] c.g.s.janitorr.JanitorrApplicationKt     : Starting JanitorrApplicationKt using Java 22.0.2 with PID 1 (/app/classes started by root in /)
2024-10-13T16:04:53.753Z  INFO 1 --- [           main] c.g.s.janitorr.JanitorrApplicationKt     : No active profile set, falling back to 1 default profile: "default"
2024-10-13T16:04:54.932Z  INFO 1 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port 8080 (http)
2024-10-13T16:04:54.941Z  INFO 1 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
2024-10-13T16:04:54.941Z  INFO 1 --- [           main] o.apache.catalina.core.StandardEngine    : Starting Servlet engine: [Apache Tomcat/10.1.30]
2024-10-13T16:04:54.967Z  INFO 1 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
2024-10-13T16:04:54.968Z  INFO 1 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 1124 ms
2024-10-13T16:04:55.260Z  WARN 1 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'mediaCleanupSchedule' defined in file [/app/classes/com/github/schaka/janitorr/cleanup/MediaCleanupSchedule.class]: Unsatisfied dependency expressed through constructor parameter 0: Error creating bean with name 'mediaServer' defined in class path resource [com/github/schaka/janitorr/mediaserver/config/MediaServerConfig.class]: Unsatisfied dependency expressed through method 'mediaServer' parameter 2: Error creating bean with name 'application-com.github.schaka.janitorr.config.ApplicationProperties': Could not bind properties to 'ApplicationProperties' : prefix=application, ignoreInvalidFields=false, ignoreUnknownFields=true
2024-10-13T16:04:55.266Z  INFO 1 --- [           main] o.apache.catalina.core.StandardService   : Stopping service [Tomcat]
2024-10-13T16:04:55.298Z  INFO 1 --- [           main] .s.b.a.l.ConditionEvaluationReportLogger : 
Error starting ApplicationContext. To display the condition evaluation report re-run your application with 'debug' enabled.
2024-10-13T16:04:55.309Z ERROR 1 --- [           main] o.s.b.d.LoggingFailureAnalysisReporter   : 
***************************
APPLICATION FAILED TO START
***************************
Description:
Failed to bind properties under 'application' to com.github.schaka.janitorr.config.ApplicationProperties:
    Reason: java.lang.NullPointerException: Parameter specified as non-null is null: method com.github.schaka.janitorr.config.ApplicationProperties.<init>, parameter mediaDeletion
Action:
Update your application's configuration
```

**Comment by Schaka** [comment]
Have a look here. Your config yaml isn't mapped correctly. 

I don't know your setup well enough to know why, but my best guess would be a permission issue where your LXC container cannot see the file on your host at all.

Are you sure your file on host is at `/config/janitorr/config/application.yml`?

**Comment by Schaka** [comment]
I still think you're accidentally mapping it incorrectly somehow.

Since I don't know exactly how you're running Dockge or Portainer, it may make sense to add the user to the compose.yml to make sure Janitorr can read the file.

To elaborate why you're seeing a nullpointer:

Spring maps several config files onto classes. They're added in order and you can overwrite properties. Janitorr comes with some sensible defaults and some that are just enough to make some components exist knowing full well they'll never be utilized.

As an example, a Rest client to access Jellyfin is created even when Jellyfin is disabled. It just points at nothing and it doesn't matter because it's never used.

For some fields, I don't provide defaults at all. Media-deletion schedule properties are some of them.

They are added on top when you provide one of the additional files. When the extra file isn't available it can't read the property at all and can't start up.

Correct file permissions, use the same owner of the file to start the container with and it should work.

**Comment by sudonem** [comment]
I'm not familiar with Spring at all, so none of the null pointer references mean anything to me unfortunately. 

If you can suggest any other changes I can make that might better expose logging data to clarify whether or not spring is seeing the application.yml file that would be super helpful. Adding the log level snippet to the application.yml file didn't change the console output in any notable way fortunately. 

I'm not new to linux or docker, but I definitely acknowledge that I could be missing something obvious here, but the fact that I've assigned the same permissions for this container as the other currently functional containers on both this docker server, and the previous I attempted, has me skeptical that the issue is permissions related.

### Other things attempted that resulted in no behavior change.
- Relocating the application.yml file to another directory with known permissions (and updated config.yml mapping)
- Creating a new user account, assigning permissions, and specifying the UID/GID in the compose file.
- Modifying the appication.yml file to disable radarr, sonarr, and jellyseer hoping to illicit different conslole output.
- Modified the compose file to disable the healthcheck.
- Removed the direct mapping for the application.yml file from compose and instead mapped it on a directory level like so:

```
version: "3"
services:
  janitorr:
    image: ghcr.io/schaka/janitorr:stable
    container_name: janitorr
    environment:
      - PUID=1000
      - PGID=1000
    volumes:
      - /opt/stacks/janitorr/config:/workspace
      - /nfs/STREAMING_MEDIA/data:/data
    healthcheck:
      test: wget -T5 -qO- http://localhost:8081/health | grep UP || exit 1
      start_period: 30s
      interval: 5s
      retries: 3
networks: {}
```
1000 is the correct UID/GID, for the docker user assigned the rest of my *arr apps - though including/excluding that explicit assignment again results in no change to the console output.

### Files and directory assigned with same permissions as the other docker containers on this server (root)
The rest of the docker containers on this server have identical permissions and are functioning normally.
![image](https://github.com/user-attachments/assets/7dcaaed0-32ac-486f-83d0-c8fe35091668)


Thanks for taking the time to respond. I appreciate it. 

Unfortunately the lack of useful logging data to indicate what might be the actual root of the misconfiguration is pretty frustrating, so I'm rapidly approaching the point of admitting defeat on this one. I could really use the tool but I'm tired of smacking my head against the wall on this one. =/

**Comment by Schaka** [comment]
Don't map the entire directory. The way things work, there are other files inside - so you want to point directly to the yaml.

Using env variables won't work. They're often used in lsio containers but are not a docker standard. You'll need to use the user parameter in compose to directly pass your uid and gid.

Also, just to be sure, please stick to the native-stable image for now.

You're not going to get more output or even be able to change the log level for Janitorr if it's not loading your config file in the first place. I suspect permissions still.

At most, you can try just running any other debugging friendly container with the exact same mapping as the same user, open its shell and see if he contents of your yaml are correct.

**Comment by Schaka** [comment]
For some reason, I couldn't see that screenshot of your files on mobile last night.
The directory and files are CLEARLY owned by root. Log in as root as use `sudo` and then do:
`chown -R 1000:1000 /opt/stacks/*`
From your description, this should be correct:

```
version: "3"
services:
  janitorr:
    image: ghcr.io/schaka/janitorr:native-stable
    container_name: janitorr
    user: 1000:1000
    volumes:
      - /opt/stacks/janitorr/config/application.yml:/workspace/application.yml
      - /nfs/STREAMING_MEDIA/data:/data
    environment:
      # Uses https://github.com/dmikusa/tiny-health-checker supplied by paketo buildpacks
      - THC_PATH=/health
      - THC_PORT=8081
    healthcheck:
      test: [ "CMD", "/cnb/process/health-check" ]
      start_period: 30s
      interval: 5s
      retries: 3
```

**Comment by Schaka** [comment]
Was this resolved after you fixed your permission issues?

**Comment by rowanfuchs** [comment]
> Have a look here. Your config yaml isn't mapped correctly.
> 
> I don't know your setup well enough to know why, but my best guess would be a permission issue where your LXC container cannot see the file on your host at all.
> 
> Are you sure your file on host is at `/config/janitorr/config/application.yml`?

I had the same error both when mapped incorrectly and with permission issues and in both cases it's a very unclear error, it would really help us users when the error has a context of the issue which causes it. I haven't been able to resolve the permission issues either.

**Comment by Schaka** [comment]
There isn't much I can do to print a different error because it's thrown by incorrect framework initialization. 
There is no direct info that the file isn't readable. The only info I have are missing properties that should've been mapped in via the config file. 

I looked into it in the past and it would require giving up compile time guarantees and additional runtime checks. Not something I'm willing to do.

I've covered it in the wiki and link directly to it in the readme.

This is the exact compose I use on my server. The `application.yml` file in `appdata/janitorr/config` on my host system is owned by user 1000 and group 1000. The `/appdata/janitorr` folder is too.
```
services:
  janitorr:
    container_name: janitorr
    image: ghcr.io/schaka/janitorr:native-stable
    user: 1000:1000
    environment:
      # Uses https://github.com/dmikusa/tiny-health-checker supplied by paketo buildpacks
      - THC_PATH=/health
      - THC_PORT=8081
    volumes:
      - /appdata/janitorr/config/application.yml:/workspace/application.yml
      - /appdata/janitorr/logs:/logs
      - /share_media:/data 
    restart: unless-stopped
    
    healthcheck:
      test: [ "CMD", "/workspace/health-check" ]
      start_period: 30s
      interval: 5s
      retries: 10
```

If you show me exactly what you're doing, I may be able to tell you what you're doing wrong.

**Comment by rowanfuchs** [comment]
Understandable it's your app and you can weigh the advantages and disadvantages, I do recommend making some sacrifices and adding a startup service that validates parameters but that is my humble opinion. 

I followed the readme but it's stuck in an endless loop, I am using Portainer:
```
2024-10-23T11:12:04.114Z  WARN 1 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'mediaCleanupSchedule' defined in file [/app/classes/com/github/schaka/janitorr/cleanup/MediaCleanupSchedule.class]: Unsatisfied dependency expressed through constructor parameter 0: Error creating bean with name 'mediaServer' defined in class path resource [com/github/schaka/janitorr/mediaserver/config/MediaServerConfig.class]: Unsatisfied dependency expressed through method 'mediaServer' parameter 2: Error creating bean with name 'bazarrService' defined in class path resource [com/github/schaka/janitorr/servarr/bazarr/BazarrConfig.class]: Unsatisfied dependency expressed through method 'bazarrService' parameter 2: Error creating bean with name 'application-com.github.schaka.janitorr.config.ApplicationProperties': Could not bind properties to 'ApplicationProperties' : prefix=application, ignoreInvalidFields=false, ignoreUnknownFields=true

2024-10-23T11:12:04.119Z  INFO 1 --- [           main] o.apache.catalina.core.StandardService   : Stopping service [Tomcat]

2024-10-23T11:12:04.143Z  INFO 1 --- [           main] .s.b.a.l.ConditionEvaluationReportLogger : 

Error starting ApplicationContext. To display the condition evaluation report re-run your application with 'debug' enabled.

2024-10-23T11:12:04.168Z ERROR 1 --- [           main] o.s.b.d.LoggingFailureAnalysisReporter   : 

***************************

APPLICATION FAILED TO START

***************************


Description:

Failed to bind properties under 'application' to com.github.schaka.janitorr.config.ApplicationProperties:

    Reason: java.lang.NullPointerException: Parameter specified as non-null is null: method com.github.schaka.janitorr.config.ApplicationProperties.<init>, parameter mediaDeletion

Action:

Update your application's configuration
```

docker compose, all root rights
```
version: '3'

services:
  janitorr:
    container_name: janitorr
    image: ghcr.io/schaka/janitorr:stable
    restart: unless-stopped
    user: 0:0
    ports:
      - 8978:8978
    volumes:
      - /hdd/config/janitorr/application.yml:/workspace/application.yml
      - /hdd/config/janitorr/logs:/logs
      - /hdd/media:/data
    healthcheck:
      test: "wget -T5 -qO- http://localhost:8081/health | grep UP || exit 1"
      start_period: 30s
      interval: 5s
      retries: 3
    networks:
      - mumedia

networks:
  mumedia:
    external: true
```

filesystem, all root rights
```
/mnt/hdd/config/janitorr/logs/janitorr.log
/mnt/hdd/config/janitorr/application.yml
/mnt/hdd/media
/mnt/hdd/media/leaving-soon
```

copied the base sample application config and didn't change anything (in my prior setup I did and experienced the same error)
```
server:
    port: 8978
  
  logging:
    level:
      com.github.schaka: INFO # Set to debug or trace to get more info about what Janitorr is doing
    threshold:
      file: NONE # Set to TRACE to enable file logging
  
  # File system access (same mapping as Sonarr, Radarr and Jellyfin) is required to delete TV shows by season and create "Leaving Soon" collections in Jellyfin
  # Currently, Jellyfin does not support an easy way to add only a few seasons or movies to a collection, we need access to temporary symlinks
  # Additionally, checks to prevent deletion on currently still seeding media currently require file system access as well
  file-system:
    access: true
    validate-seeding: true # validates seeding by checking if the original file exists and skips deletion - turning this off will send a delete to the *arrs even if a torrent may still be active
    leaving-soon-dir: "/data/media/leaving-soon" # The directory that's known to Janitorr - this will contain new folders with symlinks to your media library for "Leaving Soon"
    media-server-leaving-soon-dir: "/data/media/leaving-soon" # This is the directory Jellyfin/Emby will be told it can find the "Leaving Soon" library, in case its mapped differently
    from-scratch: true # Clean up entire "Leaving Soon" directory and rebuild from scratch - this can help with clearing orphaned data - turning this off can save resources (less writes to drive)
    free-space-check-dir: "/" # This is the default directory Janitorr uses to check how much space is left on your drives. By default, it checks the entire root - you may point it at a specific folder
  
  application:
    dry-run: true
    run-once: false # If you enable this, Janitorr will clean up once and then shut down.
    whole-tv-show: false # activating this will treat as a whole show as recently download/watched from a single episode, rather than that episode's season - shows will be deleted as a whole
    whole-show-seeding-check: false # Turning this off, disables the seeding check entirely if whole-tv-show is enabled. Activating this check will keep a whole TV show if any season is still seeding (requires file access).
    leaving-soon: 14d # 14 days before a movie is deleted, it gets added to a "Leaving Soon" type collection (i.e. movies that are 76 to 89 days old)
    exclusion-tag: "janitorr_keep" # Set this tag to your movies or TV shows in the *arrs to exclude media from being cleaned up
  
    media-deletion:
      enabled: true
      movie-expiration:
        # Percentage of free disk space to expiration time - if the highest given number is not reached, nothing will be deleted
        # If filesystem access is not given, disk percentage can't be determined. As a result, Janitorr will always choose the largest expiration time.
        5: 15d # 15 days
        10: 30d # 1 month - if a movie's files on your system are older than this, they will be deleted
        15: 30d # 2 months
        20: 90d # 3 months
      season-expiration:
        5: 15d # 15 days
        10: 20d # 20 days - if a season's files on your system are older than this, they will be deleted
        15: 60d # 2 months
        20: 120d # 4 months
  
    tag-based-deletion:
      enabled: true
      minimum-free-disk-percent: 100
      schedules:
        - tag: 5 - demo
          expiration: 30d
        - tag: 10 - demo
          expiration: 7d
  
    episode-deletion: # This ignores Jellystat. Only grab history matters. It also doesn't clean up Jellyfin. There is NO seeding check either.
      enabled: true
      tag: janitorr_daily # Shows tagged with this will have all episodes of their LATEST season deleted by the below thresholds
      max-episodes: 10 # maximum (latest) episodes of this season to keep
      max-age: 30d # Maximum age to keep any episode at all - even the last 10 episodes would expire after 30 days in this example
  
  
  clients:
    sonarr:
      enabled: true
      url: "http://localhost:8989"
      api-key: "4ed7f4d0e8584d65aa2d47d944077ff6"
      delete-empty-shows: true # Delete empty shows if deleting by season. Otherwise leaves Sonarr entries behind.
      determine-age-by: most_recent # Optional property, use 'most_recent' or 'oldest' - remove this line if Janitorr should determine by upgrades enabled for your profile
    radarr:
      enabled: true
      url: "http://localhost:7878"
      api-key: "cd0912f129d348c9b69bb20d49fcbe44"
      only-delete-files: true # NOT RECOMMENDED - When set to true, Janitorr will only delete your media files but keep the entries in Radarr
      determine-age-by: most_recent # Optional property, use 'most_recent' or 'oldest' - remove this line if Janitorr should determine by upgrades enabled for your profile
    bazarr:
      enabled: false # Only used if you want to copy over subtitle files managed by Bazarr
      url: "http://localhost:6767"
      api-key: "cd0912f129d348c9b69bb20d49fcbe55"
  
    ## You can only choose one out of Jellyfin or Emby.
    ## User login is only needed if deletion is enabled.
    jellyfin:
      enabled: true
      url: "http://localhost:8096"
      api-key: "4da8d93992804489ba2d1f0e31b8316c"
      username: Janitorr
      password: janitorr
      delete: true # Jellyfin setup is required for JellyStat. However, if you don't want Janitorr to send delete requests to the Jellyfin API, disable it here
      leaving-soon-tv: "Shows (Leaving Soon)"
      leaving-soon-movies: "Movies (Leaving Soon)"
  
    ## You can only choose one out of Jellyfin or Emby. Emby support is secondary.
    ## User login is only needed if deletion is enabled.
    emby:
      enabled: false
      url: "http://localhost:8096"
      api-key: "4da8d93992804489ba2d1f0e31b8316c"
      username: Janitorr
      password: janitorr
      delete: true # Emby setup is required for JellyStat. However, if you don't want Janitorr to send delete requests to the Emby API, disable it here
      leaving-soon-tv: "Shows (Leaving Soon)"
      leaving-soon-movies: "Movies (Leaving Soon)"
  
    jellyseerr:
      enabled: true
      url: "http://localhost:5055"
      api-key: "MTY3NzU3NzI0NzgzOWFhNWYxMGE4LWNlMWYtNDc1ZS04ODYzLThkMjQyMTQ4M2NiZCe="
      match-server: false # Enable if you have several Radarr/Sonarr instances set up in Jellyseerr. Janitorr will match them by the host+port supplied in their respective config settings.
  
    jellystat:
      enabled: true
      whole-tv-show: false # Enabling this will make Jellystat consider TV shows as a whole if any episode of any season has been watched
      url: "http://jellystat:3000"
      api-key: "jellystat-key"
```

**Comment by rowanfuchs** [comment]
to add to my comment, this is the error after the unclear mapping issue.

**Comment by Schaka** [comment]
Is there any reason you're not using the recommended native image (native-stable)?
I would highly recommend not running everything as root and having all the files owned by it. 
There are 2 things you can try here:

Stick with native image, but map your config file `/config/application.yml` inside the container. [Both files are read](https://github.com/Schaka/janitorr/blob/main/build.gradle.kts#L245) by the image, but something else may be conflicting.

That's:
```
- /appdata/janitorr/config/application.yml:/config/application.yml
``` 

Run your container not as `user: 0:0` but instead as `privileged: true`.

I also noticed the files are `yaml` inside the image, not `yml` like in the readme. I'm going to push a fix to adress both files explicitly, even though Spring does address both and I'm going to deprecate the `*.yaml` extension from being explicitly mapped.

**Comment by rowanfuchs** [comment]
`janitorr:stable` was mentioned as first in the readme, there was no other reason. `user: 0:0` was to test if permission were the issue, even with `privileged: true` or without the same error occurs. workspace or config doesn't make a difference with the native-stable image:

```

services:
  janitorr:
    container_name: janitorr
    image: ghcr.io/schaka/janitorr:native-stable
    restart: unless-stopped
    ports:
      - 8978:8978
    volumes:
      - /hdd/config/janitorr/application.yml:/workspace/application.yml
      - /hdd/config/janitorr/logs:/logs
      - /hdd/media:/data
    healthcheck:
      test: "wget -T5 -qO- http://localhost:8081/health | grep UP || exit 1"
      start_period: 30s
      interval: 5s
      retries: 3
    networks:
      - mumedia

networks:
  mumedia:
    external: true
```

```
2024-10-23T12:20:31.352Z  INFO 1 --- [           main] c.g.s.janitorr.JanitorrApplicationKt     : Starting AOT-processed JanitorrApplicationKt using Java 23 with PID 1 (/workspace/com.github.schaka.janitorr.JanitorrApplicationKt started by cnb in /workspace)

2024-10-23T12:20:31.352Z  INFO 1 --- [           main] c.g.s.janitorr.JanitorrApplicationKt     : No active profile set, falling back to 1 default profile: "default"

2024-10-23T12:20:31.405Z  INFO 1 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port 8080 (http)

2024-10-23T12:20:31.406Z  INFO 1 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]

2024-10-23T12:20:31.406Z  INFO 1 --- [           main] o.apache.catalina.core.StandardEngine    : Starting Servlet engine: [Apache Tomcat/10.1.30]

2024-10-23T12:20:31.410Z  INFO 1 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext

2024-10-23T12:20:31.410Z  INFO 1 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 58 ms

2024-10-23T12:20:31.498Z  WARN 1 --- [           main] w.s.c.ServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'mediaCleanupSchedule': Unsatisfied dependency expressed through constructor parameter 0: Error creating bean with name 'mediaServer': Unsatisfied dependency expressed through method 'mediaServer' parameter 2: Error creating bean with name 'bazarrService': Unsatisfied dependency expressed through method 'bazarrService' parameter 2: Error creating bean with name 'application-com.github.schaka.janitorr.config.ApplicationProperties': Could not bind properties to 'ApplicationProperties' : prefix=application, ignoreInvalidFields=false, ignoreUnknownFields=true

2024-10-23T12:20:31.498Z  INFO 1 --- [           main] o.apache.catalina.core.StandardService   : Stopping service [Tomcat]

2024-10-23T12:20:31.499Z ERROR 1 --- [           main] o.s.b.d.LoggingFailureAnalysisReporter   : 

***************************

APPLICATION FAILED TO START

***************************


Description:

Failed to bind properties under 'application' to com.github.schaka.janitorr.config.ApplicationProperties:

    Reason: java.lang.NullPointerException: Parameter specified as non-null is null: method com.github.schaka.janitorr.config.ApplicationProperties.<init>, parameter mediaDeletion

Action:

Update your application's configuration
```

**Comment by rowanfuchs** [comment]
My apologies, that was indeed a mistake when I was replicating the issue, with the correct mapping. I'm not sure what helped it afterwards as I started it and it still failed, I again removed the existing image manually, replaced CRLF with LF linebreaks, copied the application template again.

And it started working, now I am trying to replicate it with all those steps but it won't fail. All I can assume is that indeed the incorrect mapping or an incorrect application.yml is at fault here.

**Comment by Schaka** [comment]
No worries, it was likely a combination of incorrect permissions/mapping and maybe the JVM image (as in, non-native).
If there was a bug with the JVM image, that's now fixed in v1.6.6 anyway. I did so blindly, as I couldn't reproduce your issue, but if there was an issue for some people it should be fixed now.

Most importantly, it's working for you.
